File: backend/routers/destinations.py
from fastapi import APIRouter, HTTPException, Query, Depends
from typing import Optional, List
import uuid
from datetime import datetime

from models import DestinationCreate, DestinationResponse
from database import get_database
from services.weather_service import weather_service
from auth import get_current_user_optional

router = APIRouter(prefix="/api/destinations", tags=["destinations"])

@router.get("/", response_model=List[DestinationResponse])
async def get_destinations(
    search: Optional[str] = None,
    country: Optional[str] = None,
    category: Optional[str] = None,
    activity: Optional[str] = None,
    featured_only: Optional[bool] = False,
    limit: Optional[int] = 100
):
    """Get destinations with optional filtering"""
    db = get_database()
    query = {}
    
    if search:
        query["$or"] = [
            {"name": {"$regex": search, "$options": "i"}},
            {"description": {"$regex": search, "$options": "i"}},
            {"city": {"$regex": search, "$options": "i"}},
            {"country": {"$regex": search, "$options": "i"}}
        ]
    
    if country:
        query["country"] = {"$regex": country, "$options": "i"}
    
    if category:
        query["category"] = {"$regex": category, "$options": "i"}
    
    if activity:
        query["activities"] = {"$in": [activity]}
    
    if featured_only:
        query["featured"] = True
    
    destinations = await db.destinations.find(query).limit(limit).to_list(length=limit)
    
    # Add weather data for each destination
    result = []
    for destination in destinations:
        destination["_id"] = str(destination["_id"])
        
        # Add weather data
        weather_data = await weather_service.get_current_weather(
            destination["city"], 
            destination["country"]
        )
        destination["weather"] = weather_data
        
        result.append(DestinationResponse(**destination))
    
    return result

@router.get("/featured", response_model=List[DestinationResponse])
async def get_featured_destinations():
    """Get featured destinations"""
    return await get_destinations(featured_only=True, limit=6)

@router.get("/countries")
async def get_countries():
    """Get all available countries"""
    db = get_database()
    countries = await db.destinations.distinct("country")
    return countries

@router.get("/activities")
async def get_activities():
    """Get all available activities"""
    db = get_database()
    activities = await db.destinations.distinct("activities")
    return activities

@router.get("/{destination_id}", response_model=DestinationResponse)
async def get_destination(destination_id: str):
    """Get specific destination by ID"""
    db = get_database()
    destination = await db.destinations.find_one({"destination_id": destination_id})
    
    if not destination:
        raise HTTPException(status_code=404, detail="Destination not found")
    
    destination["_id"] = str(destination["_id"])
    
    # Add weather data
    weather_data = await weather_service.get_current_weather(
        destination["city"], 
        destination["country"]
    )
    destination["weather"] = weather_data
    
    return DestinationResponse(**destination)

@router.post("/", response_model=DestinationResponse)
async def create_destination(
    destination: DestinationCreate,
    current_user: dict = Depends(get_current_user_optional)
):
    """Create a new destination (admin only for now)"""
    db = get_database()
    
    destination_data = destination.dict()
    destination_data["destination_id"] = str(uuid.uuid4())
    destination_data["created_at"] = datetime.utcnow()
    
    result = await db.destinations.insert_one(destination_data)
    destination_data["_id"] = str(result.inserted_id)
    
    return DestinationResponse(**destination_data)

File: backend/routes/destinations.js
const express = require('express');
const { body, validationResult, query } = require('express-validator');
const Destination = require('../models/Destination');
const Review = require('../models/Review');
const auth = require('../middleware/auth');
const axios = require('axios');

const router = express.Router();

// Weather service function
const getWeatherData = async (city) => {
  if (!process.env.WEATHER_API_KEY || process.env.WEATHER_API_KEY === 'your-openweathermap-api-key-here') {
    return { 
      error: 'Weather API key not configured',
      temperature: 'N/A',
      description: 'Weather data unavailable'
    };
  }
  
  try {
    const url = `http://api.openweathermap.org/data/2.5/weather?q=${city}&appid=${process.env.WEATHER_API_KEY}&units=metric`;
    const response = await axios.get(url);
    
    if (response.status === 200) {
      return {
        temperature: response.data.main.temp,
        description: response.data.weather[0].description,
        humidity: response.data.main.humidity,
        wind_speed: response.data.wind.speed
      };
    } else {
      return { error: 'Weather data not available' };
    }
  } catch (error) {
    return { 
      error: 'Weather service unavailable',
      temperature: 'N/A',
      description: 'Weather data unavailable'
    };
  }
};

// @route   GET /api/destinations
// @desc    Get all destinations with filters
// @access  Public
router.get('/', [
  query('search').optional().isString(),
  query('country').optional().isString(),
  query('category').optional().isString(),
  query('activity').optional().isString(),
  query('featured_only').optional().isBoolean(),
  query('limit').optional().isInt({ min: 1, max: 100 })
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { search, country, category, activity, featured_only, limit = 100 } = req.query;
    
    let query = {};
    
    // Search filter
    if (search) {
      query.$or = [
        { name: { $regex: search, $options: 'i' } },
        { description: { $regex: search, $options: 'i' } },
        { city: { $regex: search, $options: 'i' } },
        { country: { $regex: search, $options: 'i' } }
      ];
    }
    
    // Country filter
    if (country) {
      query.country = { $regex: country, $options: 'i' };
    }
    
    // Category filter
    if (category) {
      query.category = { $regex: category, $options: 'i' };
    }
    
    // Activity filter
    if (activity) {
      query.activities = { $in: [activity] };
    }
    
    // Featured only filter
    if (featured_only === 'true') {
      query.featured = true;
    }
    
    const destinations = await Destination.find(query).limit(parseInt(limit));
    
    // Add weather data for each destination
    const destinationsWithWeather = await Promise.all(
      destinations.map(async (destination) => {
        const weatherData = await getWeatherData(destination.city);
        return {
          ...destination.toObject(),
          weather: weatherData
        };
      })
    );
    
    res.json(destinationsWithWeather);
  } catch (error) {
    console.error('Get destinations error:', error);
    res.status(500).json({ error: 'Server error fetching destinations' });
  }
});

// @route   GET /api/destinations/countries
// @desc    Get all available countries
// @access  Public
router.get('/countries', async (req, res) => {
  try {
    const countries = await Destination.distinct('country');
    res.json(countries);
  } catch (error) {
    console.error('Get countries error:', error);
    res.status(500).json({ error: 'Server error fetching countries' });
  }
});

// @route   GET /api/destinations/activities
// @desc    Get all available activities
// @access  Public
router.get('/activities', async (req, res) => {
  try {
    const activities = await Destination.distinct('activities');
    res.json(activities);
  } catch (error) {
    console.error('Get activities error:', error);
    res.status(500).json({ error: 'Server error fetching activities' });
  }
});

// @route   GET /api/destinations/featured
// @desc    Get featured destinations
// @access  Public
router.get('/featured', async (req, res) => {
  try {
    const destinations = await Destination.find({ featured: true }).limit(6);
    
    // Add weather data for each destination
    const destinationsWithWeather = await Promise.all(
      destinations.map(async (destination) => {
        const weatherData = await getWeatherData(destination.city);
        return {
          ...destination.toObject(),
          weather: weatherData
        };
      })
    );
    
    res.json(destinationsWithWeather);
  } catch (error) {
    console.error('Get featured destinations error:', error);
    res.status(500).json({ error: 'Server error fetching featured destinations' });
  }
});

// @route   GET /api/destinations/categories
// @desc    Get destination categories with counts
// @access  Public
router.get('/categories', async (req, res) => {
  try {
    const categories = await Destination.aggregate([
      { $group: { _id: '$category', count: { $sum: 1 } } },
      { $sort: { count: -1 } }
    ]);
    
    const formattedCategories = categories.map(cat => ({
      name: cat._id,
      count: cat.count
    }));
    
    res.json(formattedCategories);
  } catch (error) {
    console.error('Get categories error:', error);
    res.status(500).json({ error: 'Server error fetching categories' });
  }
});

// @route   GET /api/destinations/by-category/:category
// @desc    Get destinations by category
// @access  Public
router.get('/by-category/:category', [
  query('limit').optional().isInt({ min: 1, max: 100 })
], async (req, res) => {
  try {
    const { category } = req.params;
    const { limit = 20 } = req.query;
    
    const destinations = await Destination.find({
      category: { $regex: category, $options: 'i' }
    }).limit(parseInt(limit));
    
    // Add weather data for each destination
    const destinationsWithWeather = await Promise.all(
      destinations.map(async (destination) => {
        const weatherData = await getWeatherData(destination.city);
        return {
          ...destination.toObject(),
          weather: weatherData
        };
      })
    );
    
    res.json(destinationsWithWeather);
  } catch (error) {
    console.error('Get destinations by category error:', error);
    res.status(500).json({ error: 'Server error fetching destinations by category' });
  }
});

// @route   GET /api/destinations/:destination_id
// @desc    Get single destination
// @access  Public
router.get('/:destination_id', async (req, res) => {
  try {
    const { destination_id } = req.params;
    
    const destination = await Destination.findOne({ destination_id });
    
    if (!destination) {
      return res.status(404).json({ error: 'Destination not found' });
    }
    
    // Add weather data
    const weatherData = await getWeatherData(destination.city);
    
    res.json({
      ...destination.toObject(),
      weather: weatherData
    });
  } catch (error) {
    console.error('Get destination error:', error);
    res.status(500).json({ error: 'Server error fetching destination' });
  }
});

// @route   POST /api/destinations
// @desc    Create new destination
// @access  Private (Admin)
router.post('/', auth, [
  body('name').isLength({ min: 1, max: 200 }).trim(),
  body('country').isLength({ min: 1, max: 100 }).trim(),
  body('city').isLength({ min: 1, max: 100 }).trim(),
  body('category').isLength({ min: 1, max: 50 }).trim(),
  body('description').isLength({ min: 1, max: 2000 }),
  body('price_range').isIn(['$', '$$', '$$$', '$$$$']),
  body('activities').isArray(),
  body('best_time_to_visit').isLength({ min: 1, max: 200 }),
  body('images').isArray().notEmpty(),
  body('coordinates.latitude').isFloat({ min: -90, max: 90 }),
  body('coordinates.longitude').isFloat({ min: -180, max: 180 })
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const destination = new Destination(req.body);
    await destination.save();
    
    res.status(201).json(destination);
  } catch (error) {
    console.error('Create destination error:', error);
    res.status(500).json({ error: 'Server error creating destination' });
  }
});

module.exports = router;